firedrake/interpolation.py

```diff
@@ -1019,10 +969,21 @@ def callable(loops, f):
 
 
 @utils.known_pyop2_safe
-def _interpolator(V, tensor, expr, subset, arguments, access, bcs=None):
+def _interpolator(tensor, expr, subset, access, bcs=None):
+    if isinstance(expr, ufl.ZeroBaseForm):
+        if access is op2.INC:
+            return ()
+        elif access is op2.WRITE:
+            return (partial(tensor.zero, subset=subset),)
+        V = expr.arguments()[-1].function_space().dual()
+        expr = interpolate(ufl.zero(V.value_shape), V)
```

Comment:
I don't understand this block. Also isn't `access` always `INC` or `WRITE`?

Reply:
The reason for `ZeroBaseForm` is that we are not allowed to have split zero-off diagonal blocks into things like `Inteprolate(0, Argument(1))`. So I had to make the form splitter and the sub-assemblers aware of zero-simplification. This is good because we are no longer generating code for things like `Function(V).inteprolate(zero())`.

Reply:
Can this be explained in a comment. This is not at all obvious by reading the code.

Reply:
We could do something like `numpy.max(x, 0, out=x)`

Reply:
But it'd be hard to get it for subset != None

Reply:
A comment explaining this whole block is fine. My issue is purely to do with readability.